### PR TITLE
Improve cancellation for HttpActionService.

### PR DESCRIPTION
**Pull request by dirong**
Already there is true way to cancel http requests.

### CODE REVIEW

File: janet-http/http-service/src/main/java/io/techery/janet/HttpActionService.java [hunk 1]
```diff
@@ -85,48 +87,79 @@ public HttpActionService(String baseUrl, HttpClient client, Converter converter)
...
+    private void putRunningAction(Object action) {
+        if (!runningRequests.containsKey(action)) {
+            runningRequests.put(action, new CopyOnWriteArrayList<Request>());
+        }
     }
 
-    private void throwIfCanceled(Object action) throws CancelException {
-        if (!runningActions.contains(action)) {
+    private void throwIfCanceled(Object action, Request request) throws CancelException {
+        if (!runningRequests.containsKey(action)
```
Comment (almozavr) [inline janet-http/http-service/src/main/java/io/techery/janet/HttpActionService.java]: containsKey and get are equal operations they do literally the same searching algorithm. So it's obsolete for perf. Should use local var instead?

Comment (dirong) [inline janet-http/http-service/src/main/java/io/techery/janet/HttpActionService.java]: yes. it's mistake.

File: janet-http/http-service/src/main/java/io/techery/janet/HttpActionService.java [hunk 2]
```diff
@@ -74,7 +75,8 @@ public HttpActionService(String baseUrl, HttpClient client, Converter converter)
         this.baseUrl = baseUrl;
         this.client = client;
         this.converter = converter;
-        this.runningActions = new CopyOnWriteArrayList<Object>();
+        this.actionHelperCache = new HashMap<Class, ActionHelper>();
+        this.runningRequests = new ConcurrentHashMap<Object, List<Request>>();
```
Comment (almozavr) [inline janet-http/http-service/src/main/java/io/techery/janet/HttpActionService.java]: Micro optimization for `ConcurrentHashMap` with initial params
https://ria101.wordpress.com/2011/12/12/concurrenthashmap-avoid-a-common-misuse/

File: janet-http/http-service/src/main/java/io/techery/janet/HttpActionService.java [hunk 3]
```diff
@@ -85,48 +87,79 @@ public HttpActionService(String baseUrl, HttpClient client, Converter converter)
...
-            throwIfCanceled(action);
-            response = client.execute(request, new ActionRequestCallback<A>(holder) {
-                private int lastProgress;
-
-                @Override public void onProgress(int progress) {
-                    if (progress > lastProgress + PROGRESS_THRESHOLD) {
-                        callback.onProgress(holder, progress);
-                        lastProgress = progress;
+            request = builder.build();
+            putRunningRequest(action, request);
```
Comment (almozavr) [inline janet-http/http-service/src/main/java/io/techery/janet/HttpActionService.java]: not sure why we putting request with action again after we've already put action (see `putRunningRequest` calls `putRunningAction`)

File: janet-http/http-service/src/main/java/io/techery/janet/HttpActionService.java [hunk 4]
```diff
@@ -85,48 +87,79 @@ public HttpActionService(String baseUrl, HttpClient client, Converter converter)
     @Override protected <A> void sendInternal(ActionHolder<A> holder) throws HttpServiceException {
         callback.onStart(holder);
         A action = holder.action();
-        runningActions.add(action);
+        putRunningAction(action);
         final ActionHelper<A> helper = getActionHelper(action.getClass());
         if (helper == null) {
             throw new JanetInternalException("Something was happened with code generator. Check dependence of janet-http-compiler");
```
Comment (almozavr) [inline janet-http/http-service/src/main/java/io/techery/janet/HttpActionService.java]: should we remove action from `runningActions` here as well?

Comment (dirong) [inline janet-http/http-service/src/main/java/io/techery/janet/HttpActionService.java]: no. we need to know when action started but doesn't have an request yet.
Also we don't know how long will the method `helper.fillRequest` be running.

File: janet-http/http-service/src/main/java/io/techery/janet/HttpActionService.java (hunk 3 above)
Comment (dirong) [inline janet-http/http-service/src/main/java/io/techery/janet/HttpActionService.java]: see the method `cancel`
we need to cancel created request using `HttpClient`

File: janet-http/http-service/src/main/java/io/techery/janet/HttpActionService.java (hunk 2 above)
Comment (dirong) [inline janet-http/http-service/src/main/java/io/techery/janet/HttpActionService.java]: yep. ok, I'm going to configure it.  
but it won't be large cache with many locks.

File: janet-http/http-service/src/main/java/io/techery/janet/HttpActionService.java (hunk 4 above)
Comment (almozavr) [inline janet-http/http-service/src/main/java/io/techery/janet/HttpActionService.java]: just call `putRunningAction` after helper check

**Pull request merged**